src/core/utilities/cellmap.hpp

```diff
@@ -0,0 +1,580 @@
+
+#ifndef PHARE_CELLMAP_H
+#define PHARE_CELLMAP_H
+#include <cstddef>
+#include <string>
+#include <iterator>
+#include <array>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+#include <algorithm>
+#include <numeric>
+#include <optional>
+
+#include "core/utilities/box/box.hpp"
+#include "core/utilities/bucketlist.hpp"
+#include "core/logger.hpp"
+#include "core/utilities/meta/meta_utilities.hpp"
+#include "core/utilities/range/range.hpp"
+
+
+namespace PHARE::core
+{
+template<std::size_t dim, std::size_t bucket_size, typename cell_index_t = int,
+         typename key_t = std::array<cell_index_t, dim>>
+class CellMap
+{
+private:
+    using bucketlist_t = BucketList<bucket_size>;
+    using cell_t       = std::array<cell_index_t, dim>;
+    using box_t        = Box<cell_index_t, dim>;
+
+
+public:
+    CellMap() = default;
+
+
+    auto nbr_cells() const { return bucketsLists_.size(); }
+
+
+
+    // add a single index to the cellmap with the specified cell
+    template<typename CellIndex>
+    void addToCell(CellIndex const& cell, std::size_t itemIndex);
+
+    static auto constexpr default_extractor = [](auto const& item) -> auto& { return item.iCell; };
+    using DefaultExtractor                  = decltype(default_extractor);
+
+
+    // same as above but cell is found with the CellExtractor
+    template<typename Array, typename CellExtractor = DefaultExtractor,
+             typename = std::enable_if_t<is_iterable_v<Array>>>
+    void add(Array const& items, std::size_t itemIndex, CellExtractor extract = default_extractor);
+
+
+
+    // add all given items indexes to the cellmap.
+    // items will be added to the cell they are in, given the CellExtractor
+    template<typename Array, typename CellExtractor = DefaultExtractor,
+             typename = std::enable_if_t<is_iterable_v<Array>, void>>
+    void add(Array const& items, CellExtractor extract = default_extractor);
+
+
+
+    // same as above but for indexes within the given range
+    template<typename Array, typename CellExtractor = DefaultExtractor,
+             typename = std::enable_if_t<is_iterable_v<Array>, void>>
+    void add(Array const& items, std::size_t first, std::size_t last,
+             CellExtractor extract = default_extractor);
+
+
+    template<typename Cells>
+    void reserve(Cells const& cells);
+
+    // number of indexes stored in that cell of the cellmap
+    /*
+    std::size_t size(cell_t cell) const
```

Comment:
TORM?

Reply:
I've put it back in, as it will get useful I think for computing the load balancing (which is basically calling that function per cell....)